fplll/hlll.cpp

```diff
@@ -0,0 +1,636 @@
+  // thanks to sqrt(sum_{i=k}^{i<n}R(k, i)^2) (indices must be checked)). In the prior version,
+  // since R(k, k)^2 was known, we directly used the test. However, the test was probably not as
+  // accurate as what we hope. However, since this formula is used in hplll
+  // (https://github.com/gilvillard/hplll/releases), this can maybe be retested. An example of
+  // matrices that was not HLLL reduced because of a fail in this test can be generated thanks to
+  // latticegen -randseed 122 r 300 30000.
+  // Such a test can be activate by compiling with -DMODIFIED_LOVASZ_TEST
+  //
+  // TODO: this section must be improved and investigated, i.e.:
+  //   * probably other improvement to be done
+  m.get_norm_square_b(ftmp0, k, expo0);  // ||b[k]||^2 = ftmp0 * 2^expo0
+  m.norm_square_R_row(ftmp1, k, 0, k - 1,
+                      expo1);  // sum_{i = 0}^{i < k - 1}R[k][i]^2 = ftmp1 * 2^expo1
+
+  // If this check is false, we need to reenable
+  // ftmp0.mul_2si(ftmp0, expo0 - expo1);
+  FPLLL_DEBUG_CHECK(expo0 == expo1);
+
+  ftmp1.sub(ftmp0, ftmp1);  // ||b[k]||^2 - sum_{i = 0}^{i < k - 1}R[k][i]^2 = ftmp1 * 2^expo1
+
+  expo0 = m.get_row_expo(k - 1);
+
+  // Here, delta * R(k - 1, k - 1)^2 = dR[k-1] * 2^(2*expo0). We want to compare
+  //   delta * R(k - 1, k - 1)^2 <= ||b[k]||^2 - sum_{i = 0}^{i < k - 1}R[k][i]^2
+  //   dR[k-1] * 2^(2*expo0) <= ftmp1 * 2^expo1
+  //   dR[k-1] <= ftmp1 * 2^(expo1 - 2*expo0)
+  ftmp1.mul_2si(ftmp1, expo1 - 2 * expo0);
+#else  // MODIFIED_LOVASZ_TEST
+  // Modified Lovasz test, following the comment above.
+  // FIXME: probably not maintened.
+
+  m.norm_square_R_row(ftmp1, k, k, m.get_n(),
+                      expo1);       // sum_{i = k}^{i < n}R[k][i]^2 = ftmp1 * 2^expo1
+#ifdef DEBUG
+  m.get_R(ftmp0, k, k - 1, expo0);  // R(k, k - 1) = ftmp0 * 2^expo0
+#else   // DEBUG
+  m.get_R(ftmp0, k, k - 1);  // R(k, k - 1) = ftmp0 * 2^expo0
+#endif  // DEBUG
+
+  ftmp0.mul(ftmp0, ftmp0);  // R(k, k - 1)^2 = ftmp0 * 2^(2 * expo 0)
+
+  // If this check is false, we need to reenable
+  // ftmp0.mul_2si(ftmp0, 2 * expo0 - expo1);  // 2 * expo0 since R(k, k-1)^2 = ftmp0 *
+  // (2^expo0)^2
+  FPLLL_DEBUG_CHECK(2 * expo0 == expo1);
+
+  ftmp1.add(ftmp0, ftmp1);  // sum_{i = k}^{i < n}R[k][i]^2 + R(k, k-1)^2 = ftmp1 * 2^expo1
+
+  expo0 = m.get_row_expo(k - 1);
+
+  // Here, delta * R(k - 1, k - 1)^2 = dR[k-1] * 2^(2*expo0). We want to compare
+  //   delta * R(k - 1, k - 1)^2 <= sum_{i = k}^{i < n}R[k][i]^2 + R(k, k-1)^2
+  //   dR[k-1] * 2^(2*expo0) <= ftmp1 * 2^expo1
+  //   dR[k-1] <= ftmp1 * 2^(expo1 - 2*expo0)
+  ftmp1.mul_2si(ftmp1, expo1 - 2 * expo0);
+#endif  // MODIFIED_LOVASZ_TEST
+
+  return (dR[k - 1].cmp(ftmp1) <= 0);
+}
+
+// TODO: there is maybe a problem. It seems that size_reduction is often (maybe always) stopped
+// because of reduced is false and almost never because the conditions about the norm of b[kappa]
+// are reached. This can be an issue.
```

Comment:
What is the status of this todo?

Reply:
O crap, I completely forget to investigate this TODO!

To explain a bit, the incomplete size-reduction algorithm (Algorithm 3 in [HLLL](http://perso.ens-lyon.fr/damien.stehle/downloads/HLLL.pdf)) performs several step to size-reduce b[kappa]. However, we remarked with Gilles Villard that the condition to break the do...until loop is not sufficient to guarantee that b[kappa] is weak size-reduced. We then modify the test, but now, this test seems never being reached (at least in my implementation) and we leave the `size_reduction` function almost only when b[kappa] remains unchanged, instead of leaving the loop when the progress is sufficient.

My guess is that it can be a problem coming from my implementation, or coming from the test to stop the loop, since it is very strict. I think that it can be seen as a limitation of the code, and a way to improve it. I can more or less give this explanation in the code to comment this TODO.

Reply:
Thanks, that would be good.

Reply:
Done with commit 2255ff6.